### PR TITLE
Adds support for environment variables to custom plugins

### LINKED ISSUE #1323: allow for parameters in configuration plugins
The custom plugins add greater flexibility to the argo deployments, however there are situations where additional prarameters to the plugin would be helpful.

In our case we have a helm chart that lives alongside application code. For deploying, we would like to be able to pull the chart to the deployment repo, `helm template`, and then use kustomize to render the final yaml.

To accomplish this we could use environment variables eg `CHART_GIT_REPO`, `CHART_PATH`, and `GIT_REVISION` declared in the Application yaml file that the plugin would be able to access and could be used to perform the above steps.

### PR DESCRIPTION
**Pull request by alexec**
Closes #1323.

### CODE REVIEW

File: pkg/apis/application/v1alpha1/types.go [hunk 1]
```diff
@@ -80,17 +87,21 @@ type ApplicationSource struct {
 	Directory *ApplicationSourceDirectory `json:"directory,omitempty" protobuf:"bytes,10,opt,name=directory"`
 	// ConfigManagementPlugin holds config management plugin specific options
 	Plugin *ApplicationSourcePlugin `json:"plugin,omitempty" protobuf:"bytes,11,opt,name=plugin"`
+	// Env is a list of environment variables passed to the tool
+	Env []Env `json:"env,omitempty" protobuf:"bytes,12,opt,name=env"`
```
Comment (alexmt) [inline pkg/apis/application/v1alpha1/types.go]: I think `Env` should be part of `ApplicationSourcePlugin`, not `ApplicationSource`

Comment (alexec) [reply pkg/apis/application/v1alpha1/types.go]: this is very much open for discussion

Comment (alexmt) [reply pkg/apis/application/v1alpha1/types.go]: The `ApplicationSourcePlugin`, `ApplicationSourceKustomize`,  `ApplicationSourceKsonnet` etc were created to hold type-specific fields. Env is specific only for `ApplicationSourcePlugin`, right ?

Comment (alexec) [reply pkg/apis/application/v1alpha1/types.go]: not sure, I'm just about to tabulate how each tool does "variables"

Comment (alexec) [reply pkg/apis/application/v1alpha1/types.go]: | Tool | Global Variables |
|-|-|
| Directory | None |
| Helm | `values.yaml` | 
| Jsonnet | TLAs | 
| Ksonnet | TLAs |
| Kustomize | None? |
| Plugin | Env Var? |

Comment (alexmt) [reply pkg/apis/application/v1alpha1/types.go]: Looks like it is just Plugin. We've tried to have generic "parameters" for all tools but it quickly became too overloaded

File: test/e2e/fixture/fixture.go [hunk 2]
```diff
@@ -70,6 +70,10 @@ func getKubeConfig(configPath string, overrides clientcmd.ConfigOverrides) *rest
 // creates e2e tests fixture: ensures that Application CRD is installed, creates temporal namespace, starts repo and api server,
 // configure currently available cluster.
 func init() {
+	// this enables colors and disables full timestamps
+	log.SetFormatter(&log.TextFormatter{ForceColors: true})
```
Comment (alexec) [inline test/e2e/fixture/fixture.go]: this changes the logging output to be ... in my mind ... nicer

Comment (alexec) [reply test/e2e/fixture/fixture.go]: Screen Shot 2019-07-02 at 15 24 22

File: docs/operator-manual/application.yaml [hunk 3]
```diff
@@ -63,7 +63,11 @@ spec:
 
     # plugin specific config
     plugin:
-    - name: mypluginname
+      name: mypluginname
```
Comment (alexec) [inline docs/operator-manual/application.yaml]: error in existing example here

File: pkg/apis/application/v1alpha1/types.go [hunk 4]
```diff
@@ -129,7 +157,7 @@ type HelmParameter struct {
 }
 
 func (h *ApplicationSourceHelm) IsZero() bool {
-	return (h.ReleaseName == "") && len(h.ValueFiles) == 0 && len(h.Parameters) == 0
+	return h == nil || (h.ReleaseName == "") && len(h.ValueFiles) == 0 && len(h.Parameters) == 0
```
Comment (alexec) [inline pkg/apis/application/v1alpha1/types.go]: bugs fixed here

**Review by alexmt — APPROVED** [review]
LGTM!

**Pull request merged**